### ISSUE TITLE
Understanding add_contrastive_loss()

### ISSUE THREAD

**Issue by adrian-dalessandro** [google-research/simclr#51]
Hello,

I'm having some difficulty following the logic of the function call for add_contrastive_loss() defined in objectives.py. It appears that it is being passed two latent vectors concatenated into one latent vector. 

(1) Are the latent vectors representing the latent vectors generated by ResNet for the N input images and the N augmented input images?

(2) If so, are both of these latent vectors aligned? i.e. reference the code; is element i of hidden1 and element i for hidden2 referencing augmented versions of the same image?

And if so, (3) Does the add_contrastive_loss code take care of the negative samples? At which line are negative samples handled?

Thanks!

**Comment by chentingpc** [comment]
the `hidden` passed to `add_contrastive_loss()` is a concatenation of total 2*N vectors from both augmented branches. And it is splitted into two vars `hidden1` and `hidden2` - each of them is of shape (bsz, dim) and they are "aligned". We use matrix multiplication of these two vars to create a (bsz, bsz) scoring matrix where the off-diagonal entries are considered as negative.

https://github.com/google-research/simclr/blob/01ddaf0bd692ee945dad7ff5fb07b26df1b9edbe/objective.py#L83

**Comment by jajajajaja121** [comment]
> the `hidden` passed to `add_contrastive_loss()` is a concatenation of total 2*N vectors from both augmented branches. And it is splitted into two vars `hidden1` and `hidden2` - each of them is of shape (bsz, dim) and they are "aligned". We use matrix multiplication of these two vars to create a (bsz, bsz) scoring matrix where the off-diagonal entries are considered as negative.
> 
> https://github.com/google-research/simclr/blob/01ddaf0bd692ee945dad7ff5fb07b26df1b9edbe/objective.py#L83

About this loss function, I don't know why you calculate the  loss by  the max score of a row, rather than directly calculate the loss between two transformed result from the same image?

**Comment by chentingpc** [comment]
the loss has two terms: a term between two transformed versions of the same image, and another term between all images in the mini-batch (logsumexp which is like max score each row). without the second term the representation could collapse (unless using other tricks to replace the second term).

**Comment by saitjinwon** [comment]
Hi, loss in <A Simple Framework for Contrastive Learning of Visual Representations> shows that the loss is  1/(2N) sum_k=1^N loss(2k-1, 2k) + loss(2k, 2k-1).

the sum of the loss loss(2k-1, 2k) and loss(2k, 2k-1) is divided by 1/(2N) not 1/N. However in the code of add_contrastive loss:

  loss_a = tf.losses.softmax_cross_entropy(      labels, tf.concat([logits_ab, logits_aa], 1), weights=weights)  loss_b = tf.losses.softmax_cross_entropy(      labels, tf.concat([logits_ba, logits_bb], 1), weights=weights)  loss = loss_a + loss_b
  return loss, logits_ab, labels


is seems that the loss is divided by 1/N because there is no explicit number 2.

Does this mean that the loss is scaled by 2? 

Thank you.

**Comment by chentingpc** [comment]
yes
